src/libstore/filetransfer.cc

```diff
@@ -759,12 +759,17 @@ struct curlFileTransfer : public FileTransfer
 
                 S3Helper s3Helper(profile, region, scheme, endpoint);
 
+                Activity act(*logger, lvlTalkative, actFileTransfer,
+                    fmt("downloading '%s'", request.uri),
+                    {request.uri}, request.parentAct);
+
                 // FIXME: implement ETag
                 auto s3Res = s3Helper.getObject(bucketName, key);
                 FileTransferResult res;
                 if (!s3Res.data)
                     throw FileTransferError(NotFound, "S3 object '%s' does not exist", request.uri);
                 res.data = std::move(*s3Res.data);
+                res.urls.push_back(request.uri);
```

Comment:
I assume some other code assumes `urls` to be non-empty, or what was the chain of events that leads to the segfault?

Perhaps we're missing an `assert(!urls.empty())` somewhere?

Reply:
`downloadFile()` uses `*res.urls.rbegin()` to get the final URL in the redirect chain.

I've added an assert.